project-koku/koku
# Cost 3199: custom lookup to allow nulls to be returned during tag exclusion

## Jira Ticket

[COST-3199](https://issues.redhat.com/browse/COST-3199)

## Description

Whenever excluding by tag, we were loosing the `no-{option}` in our data returned. Which was an expected behavior we would like to keep. After digging into it a bit I realized that a `IS NOT NULL` is being added at the django optimizer level. Part of what django is trying to do is handle the tri-value logic, where nulls are considered unknown
```
NOT true == false
NOT false == True
NOT unknown is still unknown
```

This [stack overflow](https://stackoverflow.com/questions/71041019/django-q-object-is-adding-extra-is-not-null) actually does a pretty good job of explaining why the is not null is added. However, the main take away is: 
```
To thus ensure that ~Q(…) is the full opposite of Q(…), it thus should add foo IS NOT NULL to the condition.
```

So, in order to bypass this, I wrote my own [django lookup](https://docs.djangoproject.com/en/3.2/howto/custom-lookups/) that will take a list of tag values and do a `NOT LIKE` on them, and then added in some django Q filters to allow emptyset values and added them to self.query_filter for when we exclude by tag. 

## Testing

1. Checkout Branch
2. Load dummy data:
```
make create-test-customer
make load-test-customer-data
```

*On main:*
```
http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[tag:app]=*&exclude[tag:app]=cost
```
Notice how `no-app` is not in the return

*On this branch*
```
http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[tag:app]=*&exclude[tag:app]=cost
http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[tag:app]=*&exclude[tag:app]=cost&exclude[tag:app]=analytics
```

notice how  `no-app` is in the return value. 

Try for all of our different tag endoings.

## Notes

...
